src/uint/boxed/shl.rs

```diff
@@ -1,108 +1,127 @@
 //! [`BoxedUint`] bitwise left shift operations.
 
-use crate::{BoxedUint, CtChoice, Limb, Word};
+use crate::{BoxedUint, Limb};
 use core::ops::{Shl, ShlAssign};
 use subtle::{Choice, ConstantTimeLess};
 
 impl BoxedUint {
     /// Computes `self << shift`.
-    /// Returns zero if `shift >= Self::BITS`.
-    pub fn shl(&self, shift: u32) -> Self {
+    /// Returns `None` if `shift >= Self::BITS`.
+    pub fn shl(&self, shift: u32) -> (Self, Choice) {
+        // `floor(log2(bits_precision - 1))` is the number of bits in the representation of `shift`
+        // (which lies in range `0 <= shift < bits_precision`).
+        let shift_bits = u32::BITS - (self.bits_precision() - 1).leading_zeros();
         let overflow = !shift.ct_lt(&self.bits_precision());
         let shift = shift % self.bits_precision();
-        let log2_bits = u32::BITS - self.bits_precision().leading_zeros();
         let mut result = self.clone();
+        let mut temp = self.clone();
 
-        for i in 0..log2_bits {
+        for i in 0..shift_bits {
             let bit = Choice::from(((shift >> i) & 1) as u8);
-            result = Self::conditional_select(&result, &result.shl_vartime(1 << i), bit);
+            temp.set_to_zero();
+            // Will not overflow by construction
+            result
+                .shl_vartime_into(&mut temp, 1 << i)
+                .expect("shift within range");
+            result.conditional_assign(&temp, bit);
         }
 
-        Self::conditional_select(
-            &result,
-            &Self::zero_with_precision(self.bits_precision()),
+        (
+            Self::conditional_select(
+                &result,
+                &Self::zero_with_precision(self.bits_precision()),
+                overflow,
+            ),
             overflow,
         )
     }
 
-    /// Computes `self << shift`.
+    /// Computes `self << shift` and writes the result into `dest`.
+    /// Returns `None` if `shift >= Self::BITS`.
+    ///
+    /// WARNING: for performance reasons, `dest` is assumed to be pre-zeroized.
     ///
     /// NOTE: this operation is variable time with respect to `shift` *ONLY*.
     ///
     /// When used with a fixed `shift`, this function is constant-time with respect to `self`.
     #[inline(always)]
-    pub fn shl_vartime(&self, shift: u32) -> Self {
-        let nlimbs = self.nlimbs();
-        let mut limbs = vec![Limb::ZERO; nlimbs].into_boxed_slice();
-
+    fn shl_vartime_into(&self, dest: &mut Self, shift: u32) -> Option<()> {
```

Comment:
As a convention I'm a fan of putting this sort of output parameter on the end, at least as far as an output buffer which is simply written into otherwise uninitialized, though I will acknowledge that's inconsistent with `*_assign` APIs, which operate on `&mut self` (and might those be applicable here?)

Reply:
It's not exactly `assign`, since it writes in a new buffer. But I don't mind changing the argument order, it's a private method anyway.